daf_butler.py

```diff
@@ -0,0 +1,942 @@
+        """Find all datasets with a particular dataset type and data ID."""
+        raise NotImplementedError()
+
+    @abstractmethod
+    def get_quantum_status(self, quantum_id: uuid.UUID) -> QuantumStatus:
+        """Get the current status or a quantum."""
+        return self.get_quantum_history(quantum_id)[-1]
+
+    @abstractmethod
+    def get_dataset_status(self, dataset_id: uuid.UUID) -> DatasetStatus:
+        """Get the current status for a dataset."""
+        return self.get_dataset_history(dataset_id)[-1]
+
+    @abstractmethod
+    def get_quantum_history(self, quantum_id: uuid.UUID) -> list[QuantumStatus]:
+        """Get the history of this quantum's status.
+
+        Current status is last.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def get_dataset_history(self, dataset_id: uuid.UUID) -> list[DatasetStatus]:
+        """Get the history of this dataset's status.
+
+        Current status is last.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def filtered(
+        self,
+        where: QuantumGraphExpression = ...,
+        bind: Mapping[str, Any] | None = None,
+        data_id: DataId | None = None,
+        **kwargs: Any,
+    ) -> QuantumGraph:
+        """Filter the quanta and datasets in this graph according to an
+        expression.
+        """
+        raise NotImplementedError()
+
+    # TODO: convenience methods for the most common graph-traversal operations.
+
+
+class WorkspaceButler(LimitedButler):
+    """Limited butler that knows about a single workspace.
+
+    Notes
+    -----
+    This is an intermediate base class that doesn't assume anything about
+    PipelineTasks or Pipelines (or even quanta, despite the act that the full
+    butler will know about quanta), making it safe to put in ``daf_butler``.
+    I'm not sure if there will be subclasses other than those that work with
+    PipelineTask - this may just be a way to avoid having to move all of
+    ``pipe_base`` into ``daf_butler`` - but maybe SPHEREx or RubinTV's
+    best-effort processing would use it.
+
+    Workspace butlers do not connect to the central SQL database by default,
+    but they do generally store the central data repository's configuration,
+    and may construct a full `Butler` internally when needed. Methods that do
+    this are always documented as such (all others are not permitted to connect
+    to the database).
+    """
+
+    def __init__(
+        self,
+        name: str,
+        input_collections: Sequence[str],
+        chain: str | None,
+        parent_config: ButlerConfig,
+        parent_write_butler: Butler | None,
+        parent_read_butler: Butler | None,
+    ):
+        self._name = name
+        self._input_collections = input_collections
+        self._chain = chain
+        self._parent_config = parent_config
+        self._parent_read_butler = parent_read_butler
+        self._parent_write_butler = parent_write_butler
```

Comment:
Now that I've started revisiting how DMTN-249 needs to be refreshed to reflect the ideas in this technote ([DM-40811](https://jira.lsstcorp.org/browse/DM-40811)), one thing I'm already pretty sure I'd do differently is how much the `WorkspaceButler` tries to anticipate the needs of `PipelineWorkspaceButler`, with all of these standard `__init__` parameters, instance attributes, and properties.  I think it'd be more useful to make `WorkspaceButler` a very minimal, no-state ABC instead, and move most of these to `PipelineWorkspaceButler` or an intermediate base class, since that opens the door to using `WorkspaceButler` as a staging are for butler imports and exports.  I'm not sure that will pan out (I'm still working through the prototyping), but it's already a sign the base `WorkspaceButler` should assume less about how it will be used.